src/d3a/d3a_core/sim_results/device_statistics.py

```diff
@@ -0,0 +1,153 @@
+
+        if trade_price_list != []:
+            _create_or_append_dict(subdict, key_name, {market.time_slot: mean(trade_price_list)})
+        else:
+            _create_or_append_dict(subdict, key_name, {market.time_slot: FILL_VALUE})
+
+        self._calc_min_max_from_sim_dict(subdict, key_name)
+
+    def _device_energy_stats(self, area: Area, subdict: Dict):
+        key_name = "trade_energy_kWh"
+        market = list(area.parent.past_markets)[-1]
+        traded_energy = 0
+        for t in market.trades:
+            if t.seller == area.name:
+                traded_energy -= t.offer.energy
+            if t.buyer == area.name:
+                traded_energy += t.offer.energy
+
+        _create_or_append_dict(subdict, key_name,
+                               {market.time_slot: traded_energy})
+
+        self._calc_min_max_from_sim_dict(subdict, key_name)
+
+    def _pv_production_stats(self, area: Area, subdict: Dict):
+        key_name = "pv_production_kWh"
+        market = list(area.parent.past_markets)[-1]
+
+        _create_or_append_dict(subdict, key_name,
+                               {market.time_slot:
+                                area.strategy.energy_production_forecast_kWh[market.time_slot]})
+
+        self._calc_min_max_from_sim_dict(subdict, key_name)
+
+    def _soc_stats(self, area: Area, subdict: Dict):
+        key_name = "soc_hist"
+        market = list(area.parent.past_markets)[-1]
+        _create_or_append_dict(subdict, key_name,
+                               {market.time_slot:
+                                area.strategy.state.charge_history[market.time_slot]})
+
+        self._calc_min_max_from_sim_dict(subdict, key_name)
+
+    def _load_profile_stats(self, area: Area, subdict: Dict):
+        key_name = "load_profile_kWh"
+        market = list(area.parent.past_markets)[-1]
+        if market.time_slot in area.strategy.energy_requirement_history_kWh:
+            _create_or_append_dict(subdict, key_name, {
+                market.time_slot: area.strategy.energy_requirement_history_kWh[market.time_slot]})
+        else:
+            _create_or_append_dict(subdict, key_name, {market.time_slot: 0})
+
+        self._calc_min_max_from_sim_dict(subdict, key_name)
+
+    def gather_device_statistics(self, area: Area, subdict: Dict):
+
+        for child in area.children:
+            if child.name not in subdict.keys():
+                subdict.update({child.name: {}})
+            if child.children == []:
+                self._gather_device_statistics(child, subdict[child.name])
+            else:
+                self.gather_device_statistics(child, subdict[child.name])
+
+    def _gather_device_statistics(self, area: Area, subdict: Dict):
+        if not hasattr(area.parent, "past_markets") or len(area.parent.past_markets) == 0:
+            return None
+
+        if isinstance(area.strategy, PVStrategy):
+
+            self._device_price_stats(area, subdict)
```

Comment:
_device_price_stats and _device_energy_stats can be moved out of the if-elif blocks to remove duplication. You can check if area.strategy is not None, then you can safely deduce that this area is a device, therefore you can call safely these methods.